src/evm.rs

```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            // validate max snapshots
+            snapshots
+                .read()
+                .map_err(|err| {
+                    log::error!("failed acquiring read lock for snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })
+                .and_then(|snapshots| {
+                    if snapshots.len() >= MAX_SNAPSHOTS as usize {
+                        log::error!("maximum number of '{}' snapshots exceeded", MAX_SNAPSHOTS);
+                        Err(into_jsrpc_error(Web3Error::InternalError))
+                    } else {
+                        Ok(())
+                    }
+                })?;
+
+            // snapshot the node
+            let snapshot = inner
+                .read()
+                .map_err(|err| {
+                    format!("failed acquiring read lock to node for snapshot: {:?}", err)
+                })
+                .and_then(|reader| reader.snapshot())
+                .map_err(|err| {
+                    log::error!("failed creating snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })?;
+            snapshots
+                .write()
+                .map(|mut snapshots| {
+                    snapshots.push(snapshot);
+                    snapshots.len()
+                })
+                .map_err(|err| {
+                    log::error!("failed storing snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })
+                .map(U64::from)
+        })
+    }
+
+    fn revert_snapshot(&self, snapshot_id: U64) -> BoxFuture<Result<bool>> {
+        let inner = Arc::clone(&self.node);
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            let mut snapshots = snapshots.write().map_err(|err| {
+                log::error!("failed acquiring read lock for snapshots: {:?}", err);
+                into_jsrpc_error(Web3Error::InternalError)
+            })?;
+            let snapshot_id_index = snapshot_id.as_usize().saturating_sub(1);
+            if snapshot_id_index >= snapshots.len() {
+                log::error!("no snapshot exists for the id '{}'", snapshot_id);
+                return Err(into_jsrpc_error(Web3Error::InternalError));
+            }
+
+            // remove all snapshots following the index and use the first snapshot for restore
+            let selected_snapshot = snapshots
+                .drain(snapshot_id_index..)
+                .next()
+                .expect("unexpected failure, value must exist");
+
+            inner
+                .write()
+                .map_err(|err| format!("failed acquiring read lock for snapshots: {:?}", err))
+                .and_then(|mut writer| writer.restore_snapshot(selected_snapshot).map(|_| true))
```

Comment:
suggestion: We should add a log statement when the node is reverting to a snapshot

```suggestion
                .and_then(|mut writer| {
                    log::info!("Reverting node to snapshot {}", snapshot_id);
                    writer.restore_snapshot(selected_snapshot).map(|_| true)
                })
```